clients/drcachesim/tests/scheduler_unit_tests.cpp

```diff
@@ -6908,6 +6909,101 @@ test_options_match()
     test.check_options();
 }
 
+// A mock noise generator that only generates TRACE_TYPE_READ records with
+// address 0xdeadbeef.
+class mock_noise_generator_t : public noise_generator_t {
+public:
+    mock_noise_generator_t() {};
+
+    mock_noise_generator_t(noise_generator_info_t &info, addr_t addr_to_generate)
+        : noise_generator_t(info)
+        , addr_to_generate_(addr_to_generate) {};
+
+protected:
+    trace_entry_t
+    generate_trace_entry() override
+    {
+        trace_entry_t generated_entry = { TRACE_TYPE_READ, 4, { addr_to_generate_ } };
+        return generated_entry;
+    }
+
+private:
+    addr_t addr_to_generate_ = 0x0;
+};
+
+static void
+test_noise_generator()
+{
+    std::cerr << "\n----------------\nTesting noise generator\n";
+    static constexpr addr_t noise_generator_addr_to_generate = 0xdeadbeef;
+    static constexpr memref_tid_t TID_A = 42;
+    static constexpr memref_tid_t TID_B = 99;
+    std::vector<trace_entry_t> refs_A = {
+        /* clang-format off */
+        make_thread(TID_A),
+        make_pid(1),
+        make_version(4),
+        make_timestamp(10),
+        make_instr(10),
+        make_timestamp(30),
+        make_instr(30),
+        make_timestamp(50),
+        make_instr(50),
+        make_exit(TID_A),
+        /* clang-format on */
+    };
+    std::vector<trace_entry_t> refs_B = {
+        /* clang-format off */
+        make_thread(TID_B),
+        make_pid(1),
+        make_version(4),
+        make_timestamp(20),
+        make_instr(20),
+        make_timestamp(40),
+        make_instr(40),
+        make_timestamp(60),
+        make_instr(60),
+        make_exit(TID_B),
+        /* clang-format on */
+    };
+    std::vector<scheduler_t::input_reader_t> readers;
+    readers.emplace_back(std::unique_ptr<mock_reader_t>(new mock_reader_t(refs_A)),
+                         std::unique_ptr<mock_reader_t>(new mock_reader_t()), TID_A);
+    readers.emplace_back(std::unique_ptr<mock_reader_t>(new mock_reader_t(refs_B)),
+                         std::unique_ptr<mock_reader_t>(new mock_reader_t()), TID_B);
+    // Add noise.
+    noise_generator_info_t noise_generator_info = { 1, 1, 10 };
+    readers.emplace_back(
+        std::unique_ptr<mock_noise_generator_t>(new mock_noise_generator_t(
+            noise_generator_info, noise_generator_addr_to_generate)),
+        std::unique_ptr<mock_noise_generator_t>(new mock_noise_generator_t()),
+        INVALID_THREAD_ID);
```

Comment:
Please add a test of using the factory.  Seems ok to change this test to use the factory instead?